esphome/components/bl0940/bl0940.cpp

```diff
@@ -7,36 +7,52 @@ namespace bl0940 {
 
 static const char *const TAG = "bl0940";
 
-static const uint8_t BL0940_READ_COMMAND = 0x50;  // 0x58 according to documentation
 static const uint8_t BL0940_FULL_PACKET = 0xAA;
-static const uint8_t BL0940_PACKET_HEADER = 0x55;  // 0x58 according to documentation
+static const uint8_t BL0940_PACKET_HEADER = 0x55;  // 0x58 according to en doc but 0x55 in cn doc
+
+static const uint8_t BL0940_READ_COMMAND = 0x58;
+static const uint8_t BL0940_WRITE_COMMAND = 0xA8;
+
+// ----------------------------------------------------
+// values from initial implementation
+
+static const uint8_t BL0940_TUYA_READ_COMMAND = 0x50;
+static const uint8_t BL0940_TUYA_WRITE_COMMAND = 0xA0;
+
+static const float BL0940_TUYA_PREF = 1430;
+static const float BL0940_TUYA_UREF = 33000;
+static const float BL0940_TUYA_IREF = 275000;
+// Measured to 297J  per click according to power consumption of 5 minutes
+// Converted to kWh (3.6MJ per kwH). Used to be 256 * 1638.4
+static const float BL0940_TUYA_EREF = 3.6e6 / 297;
+//
+// ----------------------------------------------------
 
-static const uint8_t BL0940_WRITE_COMMAND = 0xA0;  // 0xA8 according to documentation
 static const uint8_t BL0940_REG_I_FAST_RMS_CTRL = 0x10;
 static const uint8_t BL0940_REG_MODE = 0x18;
 static const uint8_t BL0940_REG_SOFT_RESET = 0x19;
 static const uint8_t BL0940_REG_USR_WRPROT = 0x1A;
 static const uint8_t BL0940_REG_TPS_CTRL = 0x1B;
 
-const uint8_t BL0940_INIT[5][6] = {
+const uint8_t BL0940_INIT[5][5] = {
     // Reset to default
-    {BL0940_WRITE_COMMAND, BL0940_REG_SOFT_RESET, 0x5A, 0x5A, 0x5A, 0x38},
+    {BL0940_REG_SOFT_RESET, 0x5A, 0x5A, 0x5A, 0x38},
     // Enable User Operation Write
-    {BL0940_WRITE_COMMAND, BL0940_REG_USR_WRPROT, 0x55, 0x00, 0x00, 0xF0},
+    {BL0940_REG_USR_WRPROT, 0x55, 0x00, 0x00, 0xF0},
     // 0x0100 = CF_UNABLE energy pulse, AC_FREQ_SEL 50Hz, RMS_UPDATE_SEL 800mS
-    {BL0940_WRITE_COMMAND, BL0940_REG_MODE, 0x00, 0x10, 0x00, 0x37},
+    {BL0940_REG_MODE, 0x00, 0x10, 0x00, 0x37},
     // 0x47FF = Over-current and leakage alarm on, Automatic temperature measurement, Interval 100mS
-    {BL0940_WRITE_COMMAND, BL0940_REG_TPS_CTRL, 0xFF, 0x47, 0x00, 0xFE},
+    {BL0940_REG_TPS_CTRL, 0xFF, 0x47, 0x00, 0xFE},
     // 0x181C = Half cycle, Fast RMS threshold 6172
-    {BL0940_WRITE_COMMAND, BL0940_REG_I_FAST_RMS_CTRL, 0x1C, 0x18, 0x00, 0x1B}};
+    {BL0940_REG_I_FAST_RMS_CTRL, 0x1C, 0x18, 0x00, 0x1B}};
 
 void BL0940::loop() {
   DataPacket buffer;
   if (!this->available()) {
     return;
   }
   if (read_array((uint8_t *) &buffer, sizeof(buffer))) {
-    if (validate_checksum(&buffer)) {
+    if (validate_checksum_(&buffer)) {
```

Comment:
Please make sure all calls such as this include the `this->` prefix. The lines above and below this one need it added, as well. 🙂 Please add it as necessary anywhere else it's missing.
```suggestion
    if (this->validate_checksum_(&buffer)) {
```

Reply:
added missing this-> pointers